aten/src/ATen/native/cpu/ScatterGatherKernel.cpp

```diff
@@ -303,7 +304,59 @@ void scatter_add_cpu_kernel(Tensor& self, int64_t dim, const Tensor& index, cons
         self_data[idx_dim * self_dim_stride] += src_data[i * src_dim_stride];
       }
     },
-      /*serial_exec=*/true);
+    /*serial_exec=*/false);
+}
+
+
+
+void scatter_reduce_cpu_kernel(Tensor& self, int64_t dim, const Tensor& index, const Tensor& src,
+                               std::string& reduce) {
+  if (index.numel() == 0) {
+    return;
+  }
+
+  dim = maybe_wrap_dim(dim, self.dim());
+  
+  scatter_shape_check(self, dim, index, src);
+  int64_t index_dim_size = ensure_nonempty_size(index, dim);
+  int64_t self_dim_size = ensure_nonempty_size(self, dim);
+  std::string method_name = "scatter_" + reduce + "_";
+
+  cpu_scatter_gather_base_kernel(
+    self, dim, index, src,
+    method_name, [&] (
+      auto* self_data, auto self_dim_stride,
+      const auto* index_data, auto index_dim_stride,
+      const auto* src_data, auto src_dim_stride
+    ) {
+    std::map<
+    std::string,
+    std::function<void(const int64_t&, const int64_t&)> > reduce_func_map = {
+      {"add", [&](const int64_t& idx_dim, const int64_t& i) {
+                self_data[idx_dim * self_dim_stride] += src_data[i * src_dim_stride];
+              }},
+      {"subtract", [&](const int64_t& idx_dim, const int64_t& i) {
+                     self_data[idx_dim * self_dim_stride] += src_data[i * src_dim_stride];
+                   }},
+      {"multiply", [&](const int64_t& idx_dim, const int64_t& i) {
+                     self_data[idx_dim * self_dim_stride] *= src_data[i * src_dim_stride];
+                   }},
+      {"divide", [&](const int64_t& idx_dim, const int64_t& i) {
+                   self_data[idx_dim * self_dim_stride] /= src_data[i * src_dim_stride];
+                 }}
+    };
```

Comment:
Better `unordered_map`. `std::string` to `const std::string`. The map can be initialized outside of the callback prior to the kernel call. Maybe you can initialize it at compile time?

Reply:
Can't initialize outside since it uses the `self_data` and `src_data` which are `auto` and cannot be declared as arguments to the lambda. Essentially doing some thing like `std::function<void(int, int, auto)>` is not possible.

Declare a type as in something like `using reduce_func = std::map<const std::string, std::function<void(const int64_t&, const int64_t&)> >`?

Reply:
`using reduce_func = std::function<...`. You can if your operation is actually a templated functor, and you only pass addressed inside, not some indexes.

Reply:
But I agree, these is this issue with the base kernel itself. You still need to create that mapping for each dispatched type. I think currently the only way to do that efficiently is inside the original kernel itself in the dispatch section.